Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -315,311 +545,498 @@ void CGContextSetFillColorSpace(CGContextRef pContext, CGColorSpaceRef colorSpac
+    FAIL_FAST_IF_FAILED(factory->CreateDrawingStateBlock(&drawingState));
+
+    context->RenderTarget()->SaveDrawingState(drawingState.Get());
+    oldState.d2dState = drawingState;
 }
 
 /**
  @Status Interoperable
 */
-void CGContextRestoreGState(CGContextRef ctx) {
-    ctx->Backing()->CGContextRestoreGState();
+void CGContextRestoreGState(CGContextRef context) {
+    if (context->GStateStack().size() <= 1) {
+        TraceError(TAG, L"Invalid attempt to pop last graphics state.");
+        return;
+    }
+
+    context->GStateStack().pop();
+    auto& newState = context->CurrentGState();
+    context->RenderTarget()->RestoreDrawingState(newState.d2dState.Get());
+    newState.d2dState = nullptr;
 }
+#pragma endregion
 
+#pragma region Shape Drawing - Operations
 /**
  @Status Interoperable
 */
-void CGContextClearRect(CGContextRef ctx, CGRect rct) {
-    if (!ctx) {
-        return;
+void CGContextClearRect(CGContextRef context, CGRect rect) {
+    ComPtr<ID2D1RenderTarget> renderTarget = context->RenderTarget();
+    renderTarget->PushAxisAlignedClip(__CGRectToD2D(rect), D2D1_ANTIALIAS_MODE_PER_PRIMITIVE);
+    renderTarget->Clear(nullptr); // transparent black clear
+    renderTarget->PopAxisAlignedClip();
+}
+
+static void __CGContextDrawGeometry(CGContextRef context, ID2D1Geometry* geometry, CGPathDrawingMode drawMode) {
+    ComPtr<ID2D1RenderTarget> renderTarget = context->RenderTarget();
+    auto& currentState = context->CurrentGState();
+
+    ComPtr<ID2D1DeviceContext> deviceContext;
+    FAIL_FAST_IF_FAILED(renderTarget.As(&deviceContext));
+
+    D2D1_SIZE_F targetSize = renderTarget->GetSize();
+    ComPtr<ID2D1Image> originalTarget;
+    deviceContext->GetTarget(&originalTarget);
+
+    deviceContext->BeginDraw();
+
+    ComPtr<ID2D1CommandList> commandList;
+    FAIL_FAST_IF_FAILED(deviceContext->CreateCommandList(&commandList));
+    deviceContext->SetTarget(commandList.Get());
+
+    CGAffineTransform& currentTransform = currentState.transform;
+    CGAffineTransform transform = CGAffineTransformScale(currentTransform, 1.0, -1.0);
+    transform = CGAffineTransformTranslate(transform, 0., targetSize.height);
+    deviceContext->SetTransform({ transform.a, transform.b, transform.c, transform.d, transform.tx, transform.ty });
+
+    if (drawMode & kCGPathFill) {
+        if (drawMode & kCGPathEOFill) {
+            // TODO(DH): GH#1077 Regenerate geometry in Even/Odd fill mode.
+        }
+        deviceContext->FillGeometry(geometry, currentState.fillBrush.Get());
+    }
+
+    if (drawMode & kCGPathStroke) {
+        ComPtr<ID2D1Factory> factory;
+        deviceContext->GetFactory(&factory);
+
+        // TODO(DH): GH#1077 Do not recreate for every drawing operation
+        ComPtr<ID2D1StrokeStyle> strokeStyle;
+        // dashes must be adjusted to be based on line width
+        std::vector<CGFloat> adjustedDashes{ currentState.dashes };
+        for (float& f : adjustedDashes) {
```

Comment:
Good idea. I'll do that when this starts to cache the stroke style.

---
In reply to: [82902120](https://github.com/Microsoft/WinObjC/pull/1126#discussion_r82902120) [](ancestors = 82902120)